internal/satellite/registry/registry.go

```diff
@@ -20,102 +20,111 @@ package registry
 import (
 	"fmt"
 	"reflect"
+	"sync"
 
 	"github.com/apache/skywalking-satellite/internal/pkg/api"
 )
 
-// The creator registry.
 // All plugins is wrote in ./plugins dir. The plugin type would be as the next level dirs,
 // such as collector, client, or queue. And the 3rd level is the plugin name, that is also
 // used as key in pluginRegistry.
-type pluginRegistry struct {
-	collectorRegistry  map[string]api.Collector
-	queueRegistry      map[string]api.Queue
-	filterRegistry     map[string]api.Filter
-	forwarderRegistry  map[string]api.Forwarder
-	parserRegistry     map[string]api.Parser
-	clientRegistry     map[string]api.Client
-	fallbackerRegistry map[string]api.Fallbacker
-}
 
 // reg is the global plugin registry
-var reg *pluginRegistry
+var reg map[reflect.Type]map[string]interface{}
+var lock sync.Mutex
 
-// Plugin type.
+// Supported plugin types
 var (
 	collectorType  = reflect.TypeOf((*api.Collector)(nil)).Elem()
 	queueType      = reflect.TypeOf((*api.Queue)(nil)).Elem()
 	filterType     = reflect.TypeOf((*api.Filter)(nil)).Elem()
-	forwardType    = reflect.TypeOf((*api.Forwarder)(nil)).Elem()
+	forwarderType  = reflect.TypeOf((*api.Forwarder)(nil)).Elem()
 	parserType     = reflect.TypeOf((*api.Parser)(nil)).Elem()
 	clientType     = reflect.TypeOf((*api.Client)(nil)).Elem()
 	fallbackerType = reflect.TypeOf((*api.Fallbacker)(nil)).Elem()
 )
 
 func init() {
-	reg = &pluginRegistry{}
-	reg.collectorRegistry = make(map[string]api.Collector)
-	reg.queueRegistry = make(map[string]api.Queue)
-	reg.filterRegistry = make(map[string]api.Filter)
-	reg.forwarderRegistry = make(map[string]api.Forwarder)
-	reg.parserRegistry = make(map[string]api.Parser)
-	reg.clientRegistry = make(map[string]api.Client)
-	reg.fallbackerRegistry = make(map[string]api.Fallbacker)
+	reg = make(map[reflect.Type]map[string]interface{})
```

Comment:
merge code above. and make sure NONE specific plugin type related code coupled with plugin management to guarantee low coupling design